python/ray/workflow/workflow_state.py

```diff
@@ -0,0 +1,290 @@
+    """The execution state of a workflow. This dataclass helps with observation
+    and debugging."""
+
+    # -------------------------------- dependencies -------------------------------- #
+
+    # The reversed dependency mapping from downstream tasks to upstream tasks.
+    in_dependency: Dict[StepID, List[StepID]] = field(default_factory=dict)
+    # The dependency mapping from upstream tasks to downstream tasks.
+    out_dependency: Dict[StepID, List[StepID]] = field(
+        default_factory=lambda: defaultdict(list)
+    )
+    # The mapping from a task to its immediate continuation.
+    next_continuation: Dict[StepID, StepID] = field(default_factory=dict)
+    # The reversed mapping from continuation to its immediate task.
+    prev_continuation: Dict[StepID, StepID] = field(default_factory=dict)
+    # The mapping from a task to its latest continuation. The task itself must not
+    # be a continuation.
+    latest_continuation: Dict[StepID, StepID] = field(default_factory=dict)
+    # The mapping from a task to the root of the continuation, i.e. the initial task
+    # that generates the lineage of continuation.
+    continuation_root: Dict[StepID, StepID] = field(default_factory=dict)
+
+    # ------------------------------- task properties ------------------------------- #
+
+    # The arguments for the task.
+    task_input_args: Dict[StepID, ray.ObjectRef] = field(default_factory=dict)
+    # The function body of the task.
+    task_func_body: Dict[StepID, Optional[Callable]] = field(default_factory=dict)
+    # The options of the task.
+    task_options: Dict[StepID, WorkflowStepRuntimeOptions] = field(default_factory=dict)
+    # The context of the task.
+    task_context: Dict[StepID, WorkflowStepContext] = field(default_factory=dict)
+    # The execution metadata of a task.
+    task_execution_metadata: Dict[StepID, TaskExecutionMetadata] = field(
+        default_factory=dict
+    )
+
+    # ------------------------------ object management ------------------------------ #
+
+    # Set of references to upstream outputs.
+    reference_set: Dict[StepID, Set[StepID]] = field(
+        default_factory=lambda: defaultdict(set)
+    )
+    # The set of pending inputs of a task. We are able to run the task
+    # when it becomes empty.
+    pending_input_set: Dict[StepID, Set[StepID]] = field(default_factory=dict)
+    # The map from a task to its in-memory outputs. It may point
+    # to its continuation, so we cannot directly represent the checkpoint of a
+    # task with its ID. This is why we use a map here.
+    output_map: Dict[StepID, WorkflowRef] = field(default_factory=dict)
+    # The map from a task to its in-storage checkpoints. It may point
+    # to its continuation, so we cannot directly represent the checkpoint of a
+    # task with its ID. This is why we use a map here.
+    checkpoint_map: Dict[StepID, WorkflowRef] = field(default_factory=dict)
+    # Outputs that are free and can be garbage collected.
```

Comment:
What does "free" mean? What does garbage collection mean? Do we clear the storage or just the metadata?

Reply:
updated